lwt/wscat.ml

```diff
@@ -1,65 +1,63 @@
 open Lwt.Infix
-open Websocket
 open Websocket_lwt_unix
 
 let section = Lwt_log.Section.make "wscat"
 
 let client uri =
-  let open Frame in
+  let open Websocket in
   Resolver_lwt.resolve_uri ~uri Resolver_lwt_unix.system
   >>= fun endp ->
   let ctx = Lazy.force Conduit_lwt_unix.default_ctx in
-  Conduit_lwt_unix.(
-    endp_to_client ~ctx endp >>= fun client -> with_connection ~ctx client uri)
-  >>= fun (recv, send) ->
-  let react fr =
-    match fr.opcode with
-    | Opcode.Ping -> send @@ Frame.create ~opcode:Opcode.Pong ()
-    | Opcode.Close ->
+  Conduit_lwt_unix.endp_to_client ~ctx endp
+  >>= fun client ->
+  connect ~ctx client uri
+  >>= fun conn ->
+  let rec react () =
+    Websocket_lwt_unix.read conn
+    >>= function
+    | {Frame.opcode= Ping; _} ->
+        write conn (Frame.create ~opcode:Pong ()) >>= react
+    | {opcode= Close; content; _} ->
         (* Immediately echo and pass this last message to the user *)
-        ( if String.length fr.content >= 2 then
-          send
-          @@ Frame.create ~opcode:Opcode.Close
-               ~content:(String.sub fr.content 0 2)
-               ()
-        else send @@ Frame.close 1000 )
-        >>= fun () -> Lwt.fail Exit
-    | Opcode.Pong -> Lwt.return_unit
-    | Opcode.Text | Opcode.Binary -> Lwt_io.printf "> %s\n> %!" fr.content
-    | _ -> send @@ Frame.close 1002 >>= fun () -> Lwt.fail Exit in
-  let rec react_forever () = recv () >>= react >>= react_forever in
+        if String.length content >= 2 then
+          write conn
+            (Frame.create ~opcode:Close ~content:(String.sub content 0 2) ())
+        else write conn (Frame.close 1000)
```

Comment:
We should only respond with a close if we didn't initiate it, and then close the channel. How about https://github.com/paurkedal/ocaml-websocket/commit/27f77f3a6939b149b4ba62a2f4d4808a18ae5e58?